crates/sui-core/src/consensus_adapter.rs

```diff
@@ -318,28 +336,37 @@ impl ConsensusAdapter {
 
     fn submit_unchecked(self: &Arc<Self>, transaction: ConsensusTransaction) -> JoinHandle<()> {
         // Reconfiguration lock is dropped when pending_consensus_transactions is persisted, before it is handled by consensus
-        let async_stage = self.clone().submit_await(transaction);
+        let async_stage = self.clone().submit_and_wait(transaction);
         // Number of this tasks is limited by `sequencing_certificate_inflight` limit
         let join_handle = spawn_monitored_task!(async_stage);
         join_handle
     }
 
     #[allow(clippy::option_map_unit_fn)]
-    async fn submit_await(self: Arc<Self>, transaction: ConsensusTransaction) {
-        let should_submit = Self::should_submit(
+    async fn submit_and_wait(self: Arc<Self>, transaction: ConsensusTransaction) {
+        let _guard = InflightDropGuard::acquire(&self);
+        let processed_waiter = self
+            .authority
+            .consensus_message_processed_notify(transaction.key())
+            .boxed();
+        let await_submit = Self::await_submit_delay(
             &self.authority.committee.load(),
             &self.authority.name,
             &transaction,
-        );
-        let _inflight_guard = if should_submit {
-            Some(InflightDropGuard::acquire(&self))
-        } else {
-            None
+        )
+        .boxed();
+        // We need to wait for some delay until we submit transaction to the consensus
+        // However, if transaction is received by consensus while we wait, we don't need to wait
+        let processed_waiter = match select(processed_waiter, await_submit).await {
+            Either::Left((processed, _await_submit)) => {
+                processed.expect("Storage error when waiting for consensus message processed");
+                None
+            }
+            Either::Right(((), processed_waiter)) => Some(processed_waiter),
         };
-        let processed_waiter = self
-            .authority
-            .consensus_message_processed_notify(transaction.key());
-        if should_submit {
+        if let Some(processed_waiter) = processed_waiter {
```

Comment:
can this `if let` just be moved into the `Either::Right` case above? Then the select doesn't need to return an option.

Reply:
It can, but I actually wanted it this way, so that we don't have huge block of code nested too deep